src/qiskit_qasm3_import/state.py

```diff
@@ -0,0 +1,233 @@
+            if (
+                isinstance(item.type, types.Gate)
+                or isinstance(
+                    item.type, (types.Int, types.Uint, types.Float, types.Angle, types.Duration)
+                )
+                and item.type.const
+            ):
+                out._global_table[name] = item  # TODO: use insert
+        return out
+
+    def local_scope_copy(self):
+        """Return a copy of the symbol table for use with a new local scope.
+
+
+        Local variables created in the target (returned) symbol table will not
+        appear in the source symbol table. Thus, these local variables are
+        discarded and not visible upon returning to the surrounding scope. In
+        contrast, changes to data referred to by global symbols will be visible
+        in the surrounding scope.
+        """
+        # pylint: disable=protected-access
+        out = SymbolTable.__new__(SymbolTable)
+        out._local_table = self._local_table.copy()
+        out._global_table = self._global_table
+        out._builtin_table = self._builtin_table
+
+        return out
+
+    def insert(self, symbol: Union[List, Symbol]):
+        """Insert a `Symbol` (or each of a list thereof) into the symbol table."""
+        if isinstance(symbol, list):
+            for sym in symbol:
+                self._insert(sym)
+        else:
+            self._insert(symbol)
+
+    def _insert(self, symbol: Symbol):  # This does not catch shadowing builtins
+        if symbol.scope == Scope.GLOBAL:
+            self._global_table[symbol.name] = symbol
+        else:
+            self._local_table[symbol.name] = symbol
+
+    def __contains__(self, name: str):
+        # Return `True` if `name` is in symbol table.
+        return (
+            name in self._builtin_table or name in self._global_table or name in self._local_table
+        )
+
+    def in_local(self, name: str):
+        """Return true if ``name`` is in the symbol table and has local scope."""
+        return name in self._local_table
+
+    def __getitem__(self, name: str):
+        # An opaque interface for looking up `name` in the symbol table.
+        for table in (self._local_table, self._global_table, self._builtin_table):
+            if symbol := table.get(name):
+                return symbol
+        raise KeyError(f"Symbol {name} not found.")
+
+    def get(self, name: str):
+        """Return `Symbol` corresponding to `name`, or `None` if none exists."""
+        # Search order determines which symbols can shadow others.
+        for table in (self._local_table, self._global_table, self._builtin_table):
+            if symbol := table.get(name):
+                return symbol
+        return None
+
+    def hardware_qubits(self):
+        """Return an iterator over the `Symbol`s referring to hardware qubits."""
+        return (
+            sym for sym in self._global_table.values() if isinstance(sym.type, types.HardwareQubit)
+        )
```

Comment:
This feels weirdly over-specific.  I'd prefer to either let/make consumers of the symbol table do the iteration themselves, or to define a general `filter` iterator that takes an arbitrary callable.

Reply:
I see your point. There are other reasons to not like this, eg, this is only used once.